source/extensions/filters/http/cache/cache_filter.cc

```diff
@@ -0,0 +1,162 @@
+
+#include "common/http/headers.h"
+
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Cache {
+
+bool CacheFilter::isCacheableRequest(Http::RequestHeaderMap& headers) {
+  const Http::HeaderEntry* method = headers.Method();
+  const Http::HeaderEntry* forwarded_proto = headers.ForwardedProto();
+  const Http::HeaderValues& header_values = Http::Headers::get();
+  // TODO(toddmgreer): Also serve HEAD requests from cache.
+  // TODO(toddmgreer): Check all the other cache-related headers.
+  return method && forwarded_proto && headers.Path() && headers.Host() &&
+         (method->value() == header_values.MethodValues.Get) &&
+         (forwarded_proto->value() == header_values.SchemeValues.Http ||
+          forwarded_proto->value() == header_values.SchemeValues.Https);
+}
+
+bool CacheFilter::isCacheableResponse(Http::ResponseHeaderMap& headers) {
+  const Http::HeaderEntry* cache_control = headers.CacheControl();
+  // TODO(toddmgreer): fully check for cacheability. See for example
+  // https://github.com/apache/incubator-pagespeed-mod/blob/master/pagespeed/kernel/http/caching_headers.h.
+  if (cache_control) {
+    return !StringUtil::caseFindToken(cache_control->value().getStringView(), ",",
+                                      Http::Headers::get().CacheControlValues.Private);
+  }
+  return false;
+}
+
+CacheFilter::CacheFilter(const envoy::extensions::filters::http::cache::v3alpha::CacheConfig&,
+                         const std::string&, Stats::Scope&, TimeSource& time_source,
+                         HttpCache& http_cache)
+    : time_source_(time_source), cache_(http_cache) {}
+
+void CacheFilter::onDestroy() {
+  lookup_ = nullptr;
+  insert_ = nullptr;
+}
+
+Http::FilterHeadersStatus CacheFilter::decodeHeaders(Http::RequestHeaderMap& headers, bool) {
+  ENVOY_STREAM_LOG(debug, "CacheFilter::decodeHeaders: {}", *decoder_callbacks_, headers);
+  if (!isCacheableRequest(headers)) {
+    ENVOY_STREAM_LOG(debug, "CacheFilter::decodeHeaders ignoring uncacheable request: {}",
+                     *decoder_callbacks_, headers);
+    return Http::FilterHeadersStatus::Continue;
+  }
+  ASSERT(decoder_callbacks_);
+  lookup_ = cache_.makeLookupContext(LookupRequest(headers, time_source_.systemTime()));
+  ASSERT(lookup_);
+
+  ENVOY_STREAM_LOG(debug, "CacheFilter::decodeHeaders starting lookup", *decoder_callbacks_);
+  lookup_->getHeaders([this](LookupResult&& result) { onHeaders(std::move(result)); });
+  if (state_ == GetHeadersState::GetHeadersResultUnusable) {
+    // onHeaders has already been called, and no usable cache entry was found--continue iteration.
+    return Http::FilterHeadersStatus::Continue;
+  }
+  // onHeaders hasn't been called yet--stop iteration to wait for it, and tell it that we stopped
+  // iteration.
+  state_ = GetHeadersState::FinishedGetHeadersCall;
+  return Http::FilterHeadersStatus::StopIteration;
```

Comment:
You probably want to use StopAllIterationAndWatermark here, otherwise you are potentially going to get data/trailers callbacks which I think you don't want? I think it's probably worthwhile for you to have an integration test. Can you add one? This will help uncover this type of issues that are hard to hit in unit tests for filters.

Reply:
Switched to StopAllIterationAndWatermark. No more alliteration here :)

You can see a slightly out of date version of the integration test in the original PR (#7198). When I split that into smaller PRs, the integration test had to slated for the last one.

Reply:
Can you just add the integration test into this PR and we can then just review it all and finish it up?

Reply:
Done.